<a href="https://colab.research.google.com/github/Teevinesh1/smart-finance-assistant-/blob/main/smart_finance_assistant.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🏦 Project Overview

Welcome to your **Smart Finance Assistant** development journey! This notebook will evolve from basic CSV processing to a complete AI-powered finance application.

**Final Application Components:**
- 💬 **AI Chat Interface** - Financial advice personality
- 📊 **Data Analysis** - CSV transaction processing  
- 🔍 **RAG System** - Retrieval from financial documents
- 🛠️ **Custom Tools** - Calculators and utilities
- 🌐 **Gradio UI** - Professional web interface

**Development Approach**: Build progressively from foundation to advanced features, using AI collaboration throughout.

---

# 🚀 Getting Started: Foundation Setup

## Initial Setup
This cell installs the necessary libraries. In a Colab environment, you would uncomment the first line.

In [1]:
# Uncomment the line below when running in Google Colab
!pip install gradio pandas hands-on-ai

# Import core libraries
import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

print("📦 Core libraries loaded successfully!")
print(f"Pandas version: {pd.__version__}")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.7/143.7 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 178.2/178.2 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.0/25.0 MB 54.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 253.0/253.0 kB 13.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 360.7/360.7 kB 19.8 MB/s eta 0:00:00
  Attempting uninstall: jiter
    Found existing installation: jiter 0.15.0
    Uninstalling jiter-0.15.0:
      Successfully uninstalled jiter-0.15.0
📦 Core libraries loaded successfully!
Pandas version: 2.2.2


## Hands-on-AI Configuration

Set up the hands-on-ai package for advanced features (chat, RAG, tools):

In [2]:
import os

# Configure hands-on-ai server connection
os.environ['HANDS_ON_AI_SERVER'] = 'https://ollama.locollm.org'
os.environ['HANDS_ON_AI_MODEL'] = 'gemma3:4b'
os.environ['HANDS_ON_AI_API_KEY'] = 'Curtin2026ISYS20015002'

print("🔑 Hands-on-AI configured successfully!")

🔑 Hands-on-AI configured successfully!


The API key this semester is:  **isys2001-assignment-key**

## Connection Test

Test that everything is working correctly:

In [3]:
from hands_on_ai.chat import get_response

# Test the connection to the hands-on-ai server
try:
    response = get_response("Hello! I'm building a Smart Finance Assistant.")
    print("✅ Hands-on-AI connection successful!")
    print(f"Response: {response}")
except Exception as e:
    print(f"❌ Connection issue: {e}")
    print("You can still work on the data processing foundation without this.")

⚙️ Switching to model: gemma3:4b ... (may take a few seconds)
✅ Hands-on-AI connection successful!
Response: That's fantastic! Building a Smart Finance Assistant is a really interesting and impactful project. I'd love to hear more about it. 

To help me understand how I can be most helpful to you, could you tell me a bit about it? For example:

*   **What's the overall goal of your assistant?** (e.g., budgeting, investment tracking, debt management, financial planning, etc.)
*   **What platforms are you targeting?** (e.g., web app, mobile app, voice assistant integration like Alexa or Google Assistant)
*   **What kind of features are you planning to include?** (e.g., transaction tracking, goal setting, reports, alerts, personalized recommendations?)
*   **What technologies are you using or considering?** (e.g., Python, JavaScript, APIs for financial institutions, machine learning?)
*   **What's your current stage of development?** (e.g., just starting, prototyping, building features?)


# 🏗️ Foundation: Data Processing Skills

Before building advanced features, establish solid data processing foundations. This section focuses on CSV transaction analysis - the core of your finance assistant.

## Foundation Skill Checkpoint ✅

**Master these basics before advancing to chat/RAG/tools:**
- [ ] Load and clean CSV transaction data
- [ ] Handle real-world data issues (dollar signs, missing values)
- [ ] Calculate spending summaries by category  
- [ ] Generate business-appropriate insights
- [ ] Format output for professional presentation
- [ ] Test functions with various data scenarios

::: {.callout-tip}
## 🤖 AI Collaboration Strategy

For this foundation work, use AI to:
1. **Generate initial code** with specific business context
2. **Handle data cleaning** and validation
3. **Create professional formatting** for outputs
4. **Suggest business insights** from data patterns
5. **Help with testing** edge cases and error handling

**Remember**: You're directing AI like a junior developer - always review and improve their suggestions!
:::

## Sample Transaction Data Setup

Create or load sample transaction data to work with:

In [4]:
# 🤖 AI Collaboration Opportunity:
# Used AI to create transaction data see in diary

# Load the april 2026 transactions
df = pd.read_csv('april_2026_transactions.csv')

# Display the first few rows to understand our data
print(" April 2026 Transactions:")
print("=" * 40)
print(df.head())

print(f"\n Dataset Info:")
print(f"Total transactions: {len(df)}")
print(f"Date range: {df['Date'].min()} to {df['Date'].max()}")
print(f"Categories: {df['Category'].unique()}")




 April 2026 Transactions:
         Date  Amount       Category                            Description
0  01/04/2026  $18.50           Food  Maccas breakfast -Bourke St Melbourne
1  02/04/2026   $4.60      Transport      Myki top-up - Flinders St Station
2  03/04/2026   $6.80           Food    Flat white & muffin - Gloria Jean's
3  04/04/2026  $24.99  Entertainment           Netflix monthly subscription
4  05/04/2026   $4.60      Transport              Myki fare - bus route 246

 Dataset Info:
Total transactions: 30
Date range: 01/04/2026 to 30/04/2026
Categories: ['Food' 'Transport' 'Entertainment' 'Clothes' 'Refund']


---

# 📊 Six-Step Development Methodology

Your notebook must demonstrate the six-step methodology with clear evidence of AI collaboration at each step.

## STEP 1: Understand the Problem

**🎯 Define Your Finance Problem**

In this section, clearly state your chosen finance problem in business terms.

::: {.callout-note}
## Problem Definition Template

*Example: "I want to help people understand their spending habits by analyzing their bank transaction exports. The system should identify spending patterns, highlight potential savings opportunities, and provide actionable financial insights in easy-to-understand language."*

**🤖 AI Prompt to Try:**
```
"Help me brainstorm specific personal finance problems that could be solved
with a CSV analysis tool. I'm interested in helping people with budgeting
and spending awareness. What are common financial challenges that data
analysis could address?"
```

**Your Problem Statement:**
I want to help people who struggle to keep stick to their budget or overspend at the beginning of the month or other period. The system should be able to identify spending patterns on a weekly basis in order to increase awareness of spending. Additionally, the system should be able to identify behavioural patterns on what categories the person spend the most.  

## STEP 2: Identify Inputs and Outputs

**📥 Define Your Data Flow**

Clearly specify what data you'll work with and what insights you'll generate.

## Input/Output Analysis Template

**Inputs:**
- CSV file with transaction data (columns: Date, Amount, Category, Description)
- User preferences (spending goals, categories to focus on)
- Time period for analysis

**Outputs:**
- Total spending summary by category
- Average spending patterns
- Spending trend analysis
- Actionable savings recommendations
- Visual spending breakdown (future enhancement)

**🤖 AI Prompt to Try:**
```
"For a personal finance assistant that analyzes CSV transaction data,
what are the most valuable inputs I should collect and outputs I should
provide? Consider both technical data processing and business value
for the user."
```

**Your Input/Output Definition:**
**Inputs:**
- CSV file with transaction data (columns: Date, Amount, Category, Description)

**Outputs:**
- Total spending by category for the month
- Average spending in comparison with weekly spending per category
- Saving recommendations

## STEP 3: Work the Problem by Hand

**✋ Manual Calculation Examples**

Show 2-3 worked examples to understand the logic before coding.


## Example Business Calculation

Given this sample data:

| Date | Amount | Category | Description |
|------|--------|----------|-------------|
| 2024-08-01 | $45.50 | Groceries | Woolworths |
| 2024-08-02 | $12.00 | Transport | Opal Card |
| 2024-08-03 | $89.95 | Entertainment | Concert |
| 2024-08-04 | -$15.00 | Refund | Returned item |

**Manual Calculations:**
- Total Spending: $45.50 + $12.00 + $89.95 - $15.00 = $132.45
- By Category: Groceries $45.50, Transport $12.00, Entertainment $89.95
- Average Transaction: $132.45 ÷ 4 = $33.11
- Insight: Entertainment represents 67% of positive spending

**🤖 AI Prompt to Try:**
```
"Give me 3 realistic examples of Australian transaction data with different
scenarios (normal spending, refunds, large purchases). Show me how to manually
calculate meaningful financial insights from each scenario."
```

**Your Manual Examples:**

Sample data:

| Date       | Amount | Category | Description |
|------------|--------|----------|-------------|
| 2026-04-01 | $18.50 | Food     | Maccas Breakfast  |
| 2026-04-02 | $4.60 | Transport |  Myki top-up   |
| 2026-04-03 | $6.80 | Food | Flat white& muffin|
| 2026-04-04 | $24.99| Entertainment | Netflix monthly subscription |
| 2026-04-05 | $4.60 | Transport | Myki fare |
| 2026-04-06 | $11.90 | Food | Subway footlong |
| 2026-04-07 | $59.99 | Clothes | Jeans - Cotton On City|


**Manual Calculations:**
- Total Spending = $18.50 + $4.60 + $6.80 + $24.99 + $4.60 + $11.90 + $59.99 = $131.38
- Spending per category: Food = $18.50 + $6.80 + $11.90 = $37.2, Transport = $4.60 + $4.60 = $9.2, Entertainment = $24.99, Clothes = $59.99
- Average Spending per weeek = Total Spending\ 7 = $131.38/7 = $18.77
- Insights on highest expenditure= ($59.99/$131.38)*100 = 45% of the money has been allocated to clothes
- Comparing total average of the month with average spending per week to see if overspent on a particular week.

## STEP 4: Write Pseudocode

**📝 Plan Your Solution Logic**

Sketch the algorithm in plain English before coding.


## Pseudocode Template
```
FUNCTION analyze_spending_data(csv_file):
    // Step 1: Load and validate data
    - Read CSV file into DataFrame
    - Check for required columns (Date, Amount, Category, Description)
    - Handle missing or invalid data
    
    // Step 2: Clean financial data  
    - Remove dollar signs from Amount column
    - Convert to numeric values
    - Separate positive spending from refunds
    
    // Step 3: Calculate business metrics
    - Group transactions by category
    - Calculate totals, averages, counts per category
    - Identify spending patterns and outliers
    
    // Step 4: Generate insights
    - Calculate percentages of total spending
    - Identify top spending categories
    - Suggest potential savings opportunities
    
    // Step 5: Format for presentation
    - Create business-friendly output
    - Include currency formatting
    - Add explanatory text for insights
    
    RETURN formatted_analysis
```

**🤖 AI Prompt to Try:**
```
"Review my pseudocode for analyzing spending data from a CSV. Are there any
edge cases I'm missing? What business logic should I add to make this more
valuable for personal finance insights?"
```


**Your Pseudocode:**
```
FUNCTION analyse_transactions_data(csv_file):
  // Step 1: Load and validate csv file
  - Read CSV file
  - check data for required columns (Categories, Date, Amount, Description)
  - handle missing value or invalid data

  // Step 2: Clean financial data
  - Remove dollar signs from amount column
  - Create separate column for refunds
  - convert to numeric value from string to compute calculations

  // Step 3: Calculate Business Metrics
  - Total transaction by category
  - Calculate average cost for each category
  - identify outliers
  - Calculate 15 day period totals and average to see when is spending more

  // Step 4: Generate Insights
  - Identify top items on which spent the most
  - Suggest whether spending more beginning or end of month
  - Suggest spending habbits to better manage budget
  - Suggest saving opportunities

  // Step 5: Format for presentation
  - Add explanatory text for insights
  - include currency formatting in output as it was removed earlier for computation
  - create business-friendly and easy to understand output

RETURN formatted_analysis

```
## STEP 5: Convert to Python

**💻 Implementation with AI Collaboration**

Now implement your solution using AI assistance. Focus on creating professional, business-appropriate code.


## 🤖 Implementation Strategy

**Effective AI Prompts for Implementation:**
```
"I'm implementing a Smart Finance Assistant. Based on my pseudocode, please create
a Python function that [specific functionality]. The code should:
- Handle real-world CSV data issues (dollar signs, missing values)
- Include clear comments explaining business logic
- Use professional variable names
- Format output for business presentation
- Include basic error handling"
```

**Remember to critique and improve AI responses before using them!**
:::

### Foundation Data Processing Functions

In [5]:
# 🤖 AI Collaboration: Data Loading and Cleaning Function
# Ask AI to help you create a robust data loading function

def load_and_clean_transaction_data(file_path):
    """
    Load and clean transaction data for the Smart Finance Assistant

    🤖 AI Collaboration Prompt:
    "Create a function to load CSV transaction data with Date, Amount, Category,
    Description columns. Handle dollar signs in Amount, missing values, and
    data validation. Include clear business-focused error messages."

    Args:
        file_path: Path to CSV file or file object
    Returns:
        pandas.DataFrame: Cleaned transaction data
    """
    # Your AI-assisted implementation goes here
    # Read the CSV into a DataFrame
    df = pd.read_csv('april_2026_transactions.csv')

    # Make all column names lowercase and strip extra spaces
    # so "Amount" and "amount" and " Amount " all work the same
    df.columns = df.columns.str.strip().str.lower()

    # Drop completely blank rows
    df = df.dropna(how="all")

    # Remove $ signs and commas from the amount column so we can do maths
    # e.g. "$1,200.00" → "1200.00"
    df["amount"] = df["amount"].str.replace(r"[$,]", "", regex=True).str.strip()

    # Convert the amount from text to a real number
    df["amount"] = pd.to_numeric(df["amount"], errors="coerce")

    # Drop any rows where the amount still couldn't be read
    df = df.dropna(subset=["amount"])

    # Parse dates into proper date objects so we can sort and filter by date
    df["date"] = pd.to_datetime(df["date"], dayfirst=True)

    # Add a "day" column with just the day number (1–30)
    # — useful later for first half vs second half comparisons
    df["day"] = df["date"].dt.day

    # Label each row as "expense" or "refund"
    # Refunds are negative numbers in our CSV
    df["type"] = df["amount"].apply(lambda x: "refund" if x < 0 else "expense")

    # Add an absolute amount column (always positive) for totals and averages
    df["amount_abs"] = df["amount"].abs()

    # Tidy up category and description text — consistent capitalisation, no stray spaces
    df["category"]    = df["category"].str.strip().str.title()
    df["description"] = df["description"].str.strip()

    # Print a quick summary so we know the file loaded correctly
    print(f"  Loaded {len(df)} transactions")
    print(f"  Expenses : {(df['type'] == 'expense').sum()}")
    print(f"  Refunds  : {(df['type'] == 'refund').sum()}")
    print()
    print(df.to_string(index=False))

    return df

    pass

# Test your function
clean_data = load_and_clean_transaction_data("april_2026_transactions.csv")
print(clean_data)

  Loaded 30 transactions
  Expenses : 27
  Refunds  : 3

      date  amount      category                              description  day    type  amount_abs
2026-04-01   18.50          Food    Maccas breakfast -Bourke St Melbourne    1 expense       18.50
2026-04-02    4.60     Transport        Myki top-up - Flinders St Station    2 expense        4.60
2026-04-03    6.80          Food      Flat white & muffin - Gloria Jean's    3 expense        6.80
2026-04-04   24.99 Entertainment             Netflix monthly subscription    4 expense       24.99
2026-04-05    4.60     Transport                Myki fare - bus route 246    5 expense        4.60
2026-04-06   11.90          Food            Subway footlong - Swanston St    6 expense       11.90
2026-04-07   59.99       Clothes                   Jeans - Cotton On City    7 expense       59.99
2026-04-08   22.50 Entertainment            Cinema ticket - Event Cinemas    8 expense       22.50
2026-04-09   42.30          Food                   W

In [6]:
# 🤖 AI Collaboration: Spending Analysis Function
# Ask AI to help you create comprehensive spending analysis

def analyze_spending_patterns(df):
    """
    Analyze spending patterns and generate business insights

    🤖 AI Collaboration Prompt:
    "Create a function that analyses spending pattern and generate  actionable financial insight.
    It should calculate total spending, total spending per category, identify outliers,
    top expensive items, calculate 2 15 day period at the start and at the end in order to see which period spending is the most,
    calculate percentage of total spending expensive items are. Use simple language and use comments"

    """
    # Your AI-assisted implementation goes here
     # Work only with expenses — exclude refunds from spending analysis
    expenses = df[df["type"] == "expense"]

    line = "─" * 55

    # ── Total spending ────────────────────────────────────────
    total_spent = expenses["amount_abs"].sum()

    print()
    print("╔" + "═" * 55 + "╗")
    print("║{:^55}║".format("  SPENDING PATTERN ANALYSIS  "))
    print("╚" + "═" * 55 + "╝")

    print(f"\n{line}")
    print("  TOTAL SPENDING")
    print(line)
    print(f"  Total spent this month : ${total_spent:,.2f}")
    print(f"  Number of purchases    : {len(expenses)}")

    # ── Spending per category ─────────────────────────────────
    # Group all transactions by category and add up the amounts
    category_totals   = expenses.groupby("category")["amount_abs"].sum()
    category_counts   = expenses.groupby("category")["amount_abs"].count()
    category_averages = expenses.groupby("category")["amount_abs"].mean()

    print(f"\n{line}")
    print("  SPENDING BY CATEGORY")
    print(line)

    # Sort from highest to lowest so the biggest categories appear first
    for cat in category_totals.sort_values(ascending=False).index:
        total = category_totals[cat]
        count = category_counts[cat]
        avg   = category_averages[cat]
        pct   = (total / total_spent) * 100   # what % of total spend is this category
        bar   = "█" * int(pct / 5)            # simple bar — each block = 5%
        print(f"  {cat:<16} ${total:>8,.2f}  {pct:>5.1f}%  "
              f"avg ${avg:,.2f}/txn  ({int(count)} txn)  {bar}")

    # ── Outlier categories ────────────────────────────────────
    # A category is an outlier if it is more than 1.5x the average category spend
    # This helps flag categories where spending is disproportionately high
    avg_category_spend = category_totals.mean()
    outliers = category_totals[category_totals > avg_category_spend * 1.5]

    print(f"\n{line}")
    print("  OUTLIER CATEGORIES  (spending > 1.5x the category average)")
    print(line)

    if outliers.empty:
        print("  No outlier categories — spending is fairly balanced.")
    else:
        for cat, total in outliers.items():
            how_many_times = total / avg_category_spend
            print(f"  ⚠  {cat} — ${total:,.2f}  "
                  f"({how_many_times:.1f}x the average of ${avg_category_spend:,.2f})")

    # ── Top 5 most expensive items ────────────────────────────
    # Sort all individual transactions by amount and take the top 5
    top_5 = expenses.nlargest(5, "amount_abs")[
        ["date", "category", "amount_abs", "description"]
    ].reset_index(drop=True)

    print(f"\n{line}")
    print("  TOP 5 MOST EXPENSIVE PURCHASES")
    print(line)

    for i, row in top_5.iterrows():
        # Show what percentage of total spending this single item represents
        pct_of_total = (row["amount_abs"] / total_spent) * 100
        print(f"  {i+1}.  {row['date'].strftime('%d %b')}  "
              f"${row['amount_abs']:>8,.2f}  ({pct_of_total:.1f}% of total)  "
              f"{row['category']:<15}  {row['description']}")

    # ── 15-day period comparison ──────────────────────────────
    # Split the month into two halves and compare spending in each
    # This shows whether spending is heavier at the start or end of the month
    period_1 = expenses[expenses["day"] <= 15]   # days 1 to 15
    period_2 = expenses[expenses["day"] >  15]   # days 16 to end of month

    period_1_total = period_1["amount_abs"].sum()
    period_2_total = period_2["amount_abs"].sum()

    # Daily average for each period — fairer comparison since both have ~15 days
    period_1_daily_avg = period_1_total / 15
    period_2_daily_avg = period_2_total / 15

    # How much more (or less) was spent in period 2 vs period 1
    difference     = period_2_total - period_1_total
    diff_direction = "more" if difference > 0 else "less"

    print(f"\n{line}")
    print("  15-DAY PERIOD COMPARISON")
    print(line)
    print(f"  Period 1  (days  1–15)  ${period_1_total:>8,.2f}  "
          f"(daily avg ${period_1_daily_avg:,.2f})")
    print(f"  Period 2  (days 16–30)  ${period_2_total:>8,.2f}  "
          f"(daily avg ${period_2_daily_avg:,.2f})")
    print(f"\n  You spent ${abs(difference):,.2f} {diff_direction} in the second half of the month.")

    # ── Actionable insights ───────────────────────────────────
    print(f"\n{line}")
    print("  INSIGHTS & RECOMMENDATIONS")
    print(line)

    # Insight 1 — biggest category
    top_cat     = category_totals.idxmax()
    top_cat_pct = (category_totals[top_cat] / total_spent) * 100
    print(f"  • {top_cat} is your biggest expense at "
          f"${category_totals[top_cat]:,.2f} ({top_cat_pct:.1f}% of total spend).")

    # Insight 2 — period spending pattern
    if period_2_total > period_1_total:
        print(f"  • Your spending increases in the second half of the month. "
              f"Consider setting a mid-month budget limit.")
    else:
        print(f"  • Your spending is higher early in the month. "
              f"Good job slowing down in the second half!")

    # Insight 3 — saving tip for discretionary (non-essential) categories
    for cat in ["Entertainment", "Clothes"]:
        if cat in category_totals:
            saving = category_totals[cat] * 0.25
            print(f"  • Cutting {cat} by 25% would save you ~${saving:,.2f} this month.")

    # Insight 4 — overall savings potential
    print(f"  • Saving 20% of your total spend = ${total_spent * 0.20:,.2f} put away this month.")

    print(f"\n{line}\n")

    pass

# Test your function
analysis = analyze_spending_patterns(clean_data)
print(analysis)


╔═══════════════════════════════════════════════════════╗
║               SPENDING PATTERN ANALYSIS               ║
╚═══════════════════════════════════════════════════════╝

───────────────────────────────────────────────────────
  TOTAL SPENDING
───────────────────────────────────────────────────────
  Total spent this month : $645.31
  Number of purchases    : 27

───────────────────────────────────────────────────────
  SPENDING BY CATEGORY
───────────────────────────────────────────────────────
  Clothes          $  253.93   39.4%  avg $63.48/txn  (4 txn)  ███████
  Food             $  244.80   37.9%  avg $22.25/txn  (11 txn)  ███████
  Entertainment    $  118.98   18.4%  avg $19.83/txn  (6 txn)  ███
  Transport        $   27.60    4.3%  avg $4.60/txn  (6 txn)  

───────────────────────────────────────────────────────
  OUTLIER CATEGORIES  (spending > 1.5x the category average)
───────────────────────────────────────────────────────
  ⚠  Clothes — $253.93  (1.6x the average of $1

In [7]:
# 🤖 AI Collaboration: Business Insights Generator
# Ask AI to help create professional financial recommendations

def generate_financial_recommendations(analysis_data):
    """
    Generate actionable financial recommendations based on spending analysis

    🤖 AI Collaboration Prompt:
    "Based on spending analysis data, create professional financial
    recommendations. Include specific savings opportunities, spending
    pattern observations, and actionable advice formatted for a
    personal finance app user."

    Args:
        analysis_data: Dictionary with spending analysis results
    Returns:
        str: Formatted recommendations report
    """
    # Your AI-assisted implementation goes here
     # Work only with expenses — exclude refunds
    expenses = analysis_data[analysis_data["type"] == "expense"]
    refunds  = analysis_data[analysis_data["type"] == "refund"]

    # Calculate the key numbers we need for recommendations
    total_spent    = expenses["amount_abs"].sum()
    total_refunds  = refunds["amount_abs"].sum()
    net_spend      = total_spent - total_refunds

    # Spending per category
    category_totals = expenses.groupby("category")["amount_abs"].sum()

    # Split into first and second half of the month
    period_1_total = expenses[expenses["day"] <= 15]["amount_abs"].sum()
    period_2_total = expenses[expenses["day"] >  15]["amount_abs"].sum()

    # Top single purchase
    top_purchase = expenses.nlargest(1, "amount_abs").iloc[0]

    line      = "─" * 60
    thin_line = "·" * 60

    print()
    print("╔" + "═" * 60 + "╗")
    print("║{:^60}║".format("  💰  PERSONAL FINANCE RECOMMENDATIONS  💰  "))
    print("╚" + "═" * 60 + "╝")

    # ── 1. Spending snapshot ──────────────────────────────────
    # Give the user a quick summary of where they stand this month
    print(f"\n{line}")
    print("  📊  YOUR MONTH AT A GLANCE")
    print(line)
    print(f"  You spent ${total_spent:,.2f} this month across {len(expenses)} purchases.")
    print(f"  After ${total_refunds:,.2f} in refunds, your net spend was ${net_spend:,.2f}.")

    # Work out the daily average so we can give a relatable benchmark
    daily_avg = total_spent / 30
    print(f"  That works out to roughly ${daily_avg:,.2f} per day.")

    # ── 2. Spending pattern observations ─────────────────────
    # Highlight what the data tells us about how the user spends
    print(f"\n{line}")
    print("  🔍  SPENDING PATTERN OBSERVATIONS")
    print(line)

    # Observation 1 — which half of the month costs more
    if period_2_total > period_1_total:
        diff = period_2_total - period_1_total
        print(f"  • End-of-month surge: You spent ${diff:,.2f} more in the second")
        print(f"    half of the month (${period_2_total:,.2f}) than the first (${period_1_total:,.2f}).")
        print(f"    This is common after payday — try setting a weekly limit")
        print(f"    to keep spending even across the month.")
    else:
        diff = period_1_total - period_2_total
        print(f"  • Strong start discipline: You spent ${diff:,.2f} more in the")
        print(f"    first half (${period_1_total:,.2f}) but slowed down in the second")
        print(f"    half (${period_2_total:,.2f}). Great spending discipline!")

    # Observation 2 — biggest category as a share of total
    top_cat       = category_totals.idxmax()
    top_cat_total = category_totals[top_cat]
    top_cat_pct   = (top_cat_total / total_spent) * 100
    print(f"\n  • {top_cat} dominates your spending at ${top_cat_total:,.2f}")
    print(f"    ({top_cat_pct:.1f}% of everything you spent this month).")

    # Observation 3 — flag if any single purchase is more than 10% of total
    top_pct = (top_purchase["amount_abs"] / total_spent) * 100
    if top_pct > 10:
        print(f"\n  • Large single purchase: {top_purchase['description']}")
        print(f"    cost ${top_purchase['amount_abs']:,.2f} — that's {top_pct:.1f}% of your")
        print(f"    entire monthly spend in one transaction.")

    # ── 3. Savings opportunities ──────────────────────────────
    # Concrete, specific actions the user can take right now
    print(f"\n{line}")
    print("  💡  SAVINGS OPPORTUNITIES")
    print(line)

    saving_number = 1   # used to number each tip

    # Tip for each discretionary (non-essential) category
    # Entertainment and Clothes are wants, not needs — easier to cut
    for cat in ["Entertainment", "Clothes"]:
        if cat in category_totals:
            cat_total    = category_totals[cat]
            save_25      = cat_total * 0.25   # saving from a 25% cut
            cat_pct      = (cat_total / total_spent) * 100
            print(f"\n  {saving_number}. Reduce {cat} spending")
            print(f"     You spent ${cat_total:,.2f} on {cat} ({cat_pct:.1f}% of total).")
            print(f"     Cutting this by 25% saves ~${save_25:,.2f}/month")
            print(f"     — that's ~${save_25 * 12:,.2f} over a full year.")
            saving_number += 1

    # Tip — build an emergency fund from net spend
    emergency_fund_contribution = net_spend * 0.10   # 10% of net spend
    print(f"\n  {saving_number}. Start an emergency fund")
    print(f"     Setting aside just 10% of your net spend (${emergency_fund_contribution:,.2f}/month)")
    print(f"     builds a ${emergency_fund_contribution * 12:,.2f} safety net over 12 months.")
    saving_number += 1

    # Tip — 50/30/20 budget check
    # 50% needs, 30% wants, 20% savings — a classic personal finance framework
    needs_budget   = net_spend * 0.50
    wants_budget   = net_spend * 0.30
    savings_budget = net_spend * 0.20
    needs_actual   = sum(category_totals.get(c, 0) for c in ["Food", "Transport"])
    wants_actual   = sum(category_totals.get(c, 0) for c in ["Entertainment", "Clothes"])

    print(f"\n  {saving_number}. Check your 50/30/20 budget balance")
    print(f"     A healthy budget splits net spend into:")
    print(f"     Needs  (50%)  target ${needs_budget:,.2f}  —  actual ${needs_actual:,.2f}")
    print(f"     Wants  (30%)  target ${wants_budget:,.2f}  —  actual ${wants_actual:,.2f}")
    print(f"     Savings(20%)  target ${savings_budget:,.2f}")

    # Flag if wants are over budget
    if wants_actual > wants_budget:
        over_by = wants_actual - wants_budget
        print(f"     ⚠  Your 'wants' spending is ${over_by:,.2f} over the recommended limit.")

    # ── 4. Actionable next steps ──────────────────────────────
    # Clear, numbered things the user can do this week
    print(f"\n{line}")
    print("  ✅  YOUR ACTION PLAN FOR NEXT MONTH")
    print(line)
    print(f"  1. Set a daily spending limit of ${daily_avg * 0.85:,.2f}")
    print(f"     (15% less than your current ${daily_avg:,.2f} daily average).")
    print(f"  2. Review your {top_cat} purchases — look for cheaper alternatives.")
    print(f"  3. Do a mid-month check-in on day 15 to stay on track.")
    print(f"  4. Automate ${savings_budget:,.2f} into savings on payday so it")
    print(f"     never gets spent.")

    print(f"\n{thin_line}")
    print(f"  Tip: Small consistent changes beat big one-off cutbacks.")
    print(f"  Even saving ${savings_budget:,.2f}/month = ${savings_budget * 12:,.2f} by this time next year.")
    print(f"{thin_line}\n")
    pass

# Test your function
analysis = clean_data
recommendations = generate_financial_recommendations(analysis)
print(analysis)



╔════════════════════════════════════════════════════════════╗
║           💰  PERSONAL FINANCE RECOMMENDATIONS  💰           ║
╚════════════════════════════════════════════════════════════╝

────────────────────────────────────────────────────────────
  📊  YOUR MONTH AT A GLANCE
────────────────────────────────────────────────────────────
  You spent $645.31 this month across 27 purchases.
  After $93.98 in refunds, your net spend was $551.33.
  That works out to roughly $21.51 per day.

────────────────────────────────────────────────────────────
  🔍  SPENDING PATTERN OBSERVATIONS
────────────────────────────────────────────────────────────
  • End-of-month surge: You spent $155.75 more in the second
    half of the month ($400.53) than the first ($244.78).
    This is common after payday — try setting a weekly limit
    to keep spending even across the month.

  • Clothes dominates your spending at $253.93
    (39.4% of everything you spent this month).

  • Large single purchase: Sn

---

# 🌐 Advanced Features: Integrating AI Components

Once your foundation data processing is solid, integrate advanced AI features using hands-on-ai.

## Chat Interface Integration

In [8]:
# 🤖 AI Collaboration: Financial Advice Chatbot
# Ask AI to help you create a finance-focused chat personality

from hands_on_ai import chat

def create_finance_chat_personality():
    """
    Set up a financial advisor chat personality using hands-on-ai

    🤖 AI Collaboration Prompt:
    "Help me create a system prompt for a friendly, professional financial
    advisor chatbot that can provide spending advice based on transaction
    analysis. The personality should be encouraging but practical."
    """
    # Your AI-assisted chat setup goes here
    # Include personality traits, knowledge focus, response style
    chat.system = (
        "You are a friendly and practical personal finance advisor called FinBot. "
        "Your personality: encouraging and supportive, never make the user feel bad about their spending, "
        "practical with specific actionable advice, simple language with no financial jargon. "
        "You know about personal budgeting, saving strategies, spending pattern analysis, "
        "the 50/30/20 budgeting rule and how to cut discretionary spending. "
        "Always end your response with one actionable tip the user can do today."
    )

# Set up the chatbot with our finance personality
create_finance_chat_personality()

# Test your chatbot
response = chat.get_response("I spend too much on coffee, what should I do?")
print(response)
pass

# Test your chatbot
# response = chat.say("I spend too much on coffee, what should I do?")
# print(response)

Okay, let's tackle this! Spending a lot on coffee is a super common problem – it's convenient and enjoyable, but costs can really add up. Here's a breakdown of what you can do, broken down into different levels of effort and commitment:

**1. Awareness & Tracking (Low Effort - Immediate Impact)**

* **Track Your Spending:** Seriously, for just a week or two, write down *every* coffee purchase. Note the price, time of day, and where you got it. This will immediately show you just how much you're really spending. There are apps like Mint, YNAB (You Need a Budget), or even just a simple spreadsheet to help with this.
* **Calculate Your Average:** Once you’ve tracked, calculate your average coffee spend per week/month. Shockingly, it's likely higher than you think!
* **Identify Your Trigger:**  Why are you buying coffee? Is it:
    * **Convenience?** Feeling rushed in the morning?
    * **Habit?** Just a routine you’ve fallen into?
    * **Social?** Meeting a friend for coffee?
    * **Tre

## RAG System for Financial Documents

In [9]:
# 🤖 AI Collaboration: Document Retrieval Setup
# Ask AI to help you set up RAG for financial documents

from hands_on_ai import rag

def setup_financial_rag():
    """
    Set up RAG system for financial documents and transaction data

    🤖 AI Collaboration Prompt:
    "Help me set up a RAG system that can retrieve information from
    financial documents, budgeting guides, and transaction summaries
    to answer user questions about personal finance."
    """
    # Your AI-assisted RAG setup goes here
    # Include document ingestion, query processing
    # Create our own finance knowledge document
    finance_document = """
    The 50/30/20 rule splits income into three parts.
    50 percent goes to needs like rent food and transport.
    30 percent goes to wants like entertainment and clothes.
    20 percent goes to savings and investments.

    To stop overspending on entertainment set a fixed monthly budget.
    Use the 24 hour rule before any non essential purchase.
    Cancel unused subscriptions and find free alternatives.

    To build savings pay yourself first before spending on anything else.
    Set up automatic transfers to savings on payday.
    Start with saving 10 percent and increase gradually.

    To reduce food spending meal prep on weekends to avoid takeaways.
    Make a shopping list before going to the store.
    Cook at home at least 5 days a week to save money.

    To manage transport costs use public transport when possible.
    Carpool with colleagues to split fuel costs.
    Walk or cycle for short distances to save on fuel.

    An emergency fund should cover 3 to 6 months of expenses.
    Keep emergency funds in a separate savings account.
    Only use emergency funds for genuine emergencies.
    """

    # Save the finance document as a text file
    with open("finance_guide.txt", "w") as f:
        f.write(finance_document)
    print("📄 Finance document created!")

    # Split the document into small chunks for searching
    chunks = rag.chunk_text(finance_document)
    sources = ["finance_guide.txt"] * len(chunks)
    print(f"✅ Document split into {len(chunks)} chunks!")

    return chunks, sources

def ask_financial_question(question, chunks, sources):
    # Search through chunks manually to find the most relevant answer
    # Since get_embeddings needs Ollama, we do simple keyword matching instead
    question_words = question.lower().split()

    # Score each chunk by how many question words it contains
    scored_chunks = []
    for chunk, source in zip(chunks, sources):
        score = sum(1 for word in question_words if word in chunk.lower())
        scored_chunks.append((score, chunk, source))

    # Sort by score and take the top 3 most relevant chunks
    scored_chunks.sort(reverse=True)
    top_chunks = scored_chunks[:3]

    print(f"\n❓ Question: {question}\n")
    print("📚 Most relevant information found:")
    print("─" * 50)
    for score, chunk, source in top_chunks:
        print(f"\n  {chunk}")
    print("─" * 50)

# Set up the RAG system
chunks, sources = setup_financial_rag()
pass

# Test your RAG system
answer = ask_financial_question(
    "What's a good budgeting strategy for someone who overspends on entertainment?",
    chunks, sources
)

📄 Finance document created!
✅ Document split into 1 chunks!

❓ Question: What's a good budgeting strategy for someone who overspends on entertainment?

📚 Most relevant information found:
──────────────────────────────────────────────────

  The 50/30/20 rule splits income into three parts. 50 percent goes to needs like rent food and transport. 30 percent goes to wants like entertainment and clothes. 20 percent goes to savings and investments. To stop overspending on entertainment set a fixed monthly budget. Use the 24 hour rule before any non essential purchase. Cancel unused subscriptions and find free alternatives. To build savings pay yourself first before spending on anything else. Set up automatic transfers to savings on payday. Start with saving 10 percent and increase gradually. To reduce food spending meal prep on weekends to avoid takeaways. Make a shopping list before going to the store. Cook at home at least 5 days a week to save money. To manage transport costs use public t

## Custom Financial Tools

In [10]:
# 🤖 AI Collaboration: Custom Tool Development
# Ask AI to help you create useful financial calculators

from hands_on_ai import agent

def create_savings_calculator_tool(current_savings: float, monthly_contribution: float, target_amount: float) -> str:
    # Savings goal calculator that takes current savings, monthly contribution
    # and target amount then calculates how long it will take to reach the goal
    # AI Collaboration Prompt: Create a savings goal calculator function that takes current savings,
    # monthly contribution, and target amount, then calculates time to reach
    # goal. Format output for user friendly display.

    # Validate inputs — all values must be positive numbers
    if current_savings < 0 or monthly_contribution <= 0 or target_amount <= 0:
        return "❌ Please enter positive values for all fields."

    # Check if goal is already reached
    if current_savings >= target_amount:
        return "✅ You have already reached your savings goal!"

    # Calculate how much more is needed
    amount_needed = target_amount - current_savings

    # Calculate how many months it will take to reach the goal
    months_needed = amount_needed / monthly_contribution

    # Convert months to years and remaining months for easier reading
    years  = int(months_needed // 12)
    months = int(months_needed % 12)

    # Format the result in a user friendly way
    line = "─" * 45
    result = (
        f"\n{line}\n"
        f"  💰 SAVINGS GOAL CALCULATOR\n"
        f"{line}\n"
        f"  Current savings     : ${current_savings:,.2f}\n"
        f"  Monthly contribution: ${monthly_contribution:,.2f}\n"
        f"  Target amount       : ${target_amount:,.2f}\n"
        f"  Amount still needed : ${amount_needed:,.2f}\n"
        f"{line}\n"
        f"  ⏱  Time to reach goal: {years} years and {months} months\n"
        f"{line}\n"
    )
    return result

# Register your tool with the agent system
agent.register_tool(
    name="savings_calculator",
    description="Calculates how long it will take to reach a savings goal based on current savings and monthly contributions",
    function=create_savings_calculator_tool
)
print("✅ Savings calculator tool registered!\n")

# Test the tool directly without the agent
result = create_savings_calculator_tool(
    current_savings=500,
    monthly_contribution=200,
    target_amount=5000
)
print(result)

# Test your tool with different savings scenarios
print("Test 1 — Standard savings goal:")
print(create_savings_calculator_tool(500, 200, 5000))

print("Test 2 — Already reached goal:")
print(create_savings_calculator_tool(6000, 200, 5000))

print("Test 3 — Large savings goal:")
print(create_savings_calculator_tool(1000, 500, 20000))

✅ Savings calculator tool registered!


─────────────────────────────────────────────
  💰 SAVINGS GOAL CALCULATOR
─────────────────────────────────────────────
  Current savings     : $500.00
  Monthly contribution: $200.00
  Target amount       : $5,000.00
  Amount still needed : $4,500.00
─────────────────────────────────────────────
  ⏱  Time to reach goal: 1 years and 10 months
─────────────────────────────────────────────

Test 1 — Standard savings goal:

─────────────────────────────────────────────
  💰 SAVINGS GOAL CALCULATOR
─────────────────────────────────────────────
  Current savings     : $500.00
  Monthly contribution: $200.00
  Target amount       : $5,000.00
  Amount still needed : $4,500.00
─────────────────────────────────────────────
  ⏱  Time to reach goal: 1 years and 10 months
─────────────────────────────────────────────

Test 2 — Already reached goal:
✅ You have already reached your savings goal!
Test 3 — Large savings goal:

────────────────────────────────────

## Gradio UI Integration

In [11]:
# 🤖 AI Collaboration: Professional UI Design
# Ask AI to help you create a comprehensive Gradio interface

import gradio as gr
from hands_on_ai import chat

def create_finance_assistant_ui():
    # Create a comprehensive Gradio interface for the Smart Finance Assistant
    # AI Collaboration Prompt: Help me design a Gradio interface that combines CSV upload,
    # spending analysis, chat functionality, and custom tools in a user friendly
    # layout suitable for a personal finance application.

    # ── Tab 1: CSV Upload and Spending Analysis ───────────────
    def run_analysis(file):
        import io
        import sys

        # Use the load_and_clean_transaction_data function to load the uploaded file
        # Capture printed output from both functions since they use print()
        captured = io.StringIO()
        sys.stdout = captured

        df = load_and_clean_transaction_data(file.name)
        analyze_spending_patterns(df)

        # Restore normal printing and return the captured output
        sys.stdout = sys.__stdout__
        return captured.getvalue()

    # ── Tab 2: Finance Chatbot ────────────────────────────────
    def chatbot_response(message, history):
        # Set up the finance personality using our existing function
        create_finance_chat_personality()

        # Get response from the chatbot using the personality we just set
        response = chat.get_response(message)
        return response

    # ── Tab 3: Savings Calculator ─────────────────────────────
    def savings_calculator(current_savings, monthly_contribution, target_amount):
        # Call our existing savings calculator tool directly
        return create_savings_calculator_tool(current_savings, monthly_contribution, target_amount)

    # ── Build the Gradio Interface ────────────────────────────
    with gr.Blocks(title="Smart Finance Assistant") as demo:

        # App title and description
        gr.Markdown("# 💰 Smart Finance Assistant")
        gr.Markdown("Your personal AI powered finance advisor — upload your transactions, chat for advice, and plan your savings.")

        # Tab 1 — Spending Analysis
        with gr.Tab("📊 Spending Analysis"):
            gr.Markdown("### Upload your CSV file to analyse your spending")
            csv_input       = gr.File(label="Upload Transactions CSV")
            analyse_btn     = gr.Button("Analyse Spending")
            analysis_output = gr.Textbox(label="Analysis Results", lines=30)
            analyse_btn.click(run_analysis, inputs=csv_input, outputs=analysis_output)

        # Tab 2 — Finance Chatbot
        with gr.Tab("🤖 Finance Chatbot"):
            gr.Markdown("### Ask FinBot anything about your finances")
            gr.ChatInterface(chatbot_response)

        # Tab 3 — Savings Calculator
        with gr.Tab("🧮 Savings Calculator"):
            gr.Markdown("### Calculate how long it will take to reach your savings goal")
            current  = gr.Number(label="Current Savings ($)", value=500)
            monthly  = gr.Number(label="Monthly Contribution ($)", value=200)
            target   = gr.Number(label="Target Amount ($)", value=5000)
            calc_btn = gr.Button("Calculate")
            calc_output = gr.Textbox(label="Result", lines=8)
            calc_btn.click(savings_calculator, inputs=[current, monthly, target], outputs=calc_output)

    return demo

# Launch your complete application
demo = create_finance_assistant_ui()
demo.launch()

It looks like you are running Gradio on a hosted Jupyter notebook, which requires `share=True`. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://752ce7d09233053b7c.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


---

# 🧪 STEP 6: Test with a Variety of Data

**🔍 Comprehensive Testing Strategy**

Create thorough tests for your Smart Finance Assistant to ensure it handles real-world scenarios.

::: {.callout-tip}
## 🤖 AI Collaboration for Testing

**Effective Testing Prompts:**
```
"Help me create comprehensive test cases for my finance assistant. Include:
- Normal transaction data
- Edge cases (refunds, large amounts, missing data)
- Invalid data scenarios (corrupted files, wrong formats)
- Business logic validation (spending calculations, recommendations)
Create assert statements to verify each scenario."
```
:::

## Foundation Function Tests

In [18]:
import pandas as pd

# =========================================================
# 1. CREATE TEST DATASETS
# =========================================================

def create_test_datasets():
    """
    Create realistic test datasets for finance assistant
    """

    # ---------------- Normal data ----------------
    normal = pd.DataFrame({
        "date": ["01/04/2026", "03/04/2026", "05/04/2026"],
        "amount": ["$45.50", "$120.00", "$250.00"],
        "category": ["Food", "Transport", "Shopping"],
        "description": ["Woolworths", "Uber", "JB Hi-Fi"]
    })

    # ---------------- Edge cases ----------------
    edge_cases = pd.DataFrame({
        "date": ["02/04/2026", "08/04/2026", "10/04/2026"],
        "amount": ["-$50.00", "$0.00", "$9999.99"],
        "category": ["Refund", "Food", "Electronics"],
        "description": ["Amazon refund", "", "MacBook Pro"]
    })

    # ---------------- Invalid data ----------------
    invalid = pd.DataFrame({
        "date": ["bad_date", "", "15/04/2026"],
        "amount": ["abc", "twenty", "$100.00"],
        "category": ["Food", None, "Transport"],
        "description": ["bad", "missing", "valid"]
    })

    # ---------------- High spending ----------------
    high_spending = pd.DataFrame({
        "date": ["01/04/2026", "10/04/2026"],
        "amount": ["$2000.00", "$1500.00"],
        "category": ["Entertainment", "Travel"],
        "description": ["Concert", "Flight"]
    })

    return {
        "normal": normal,
        "edge": edge_cases,
        "invalid": invalid,
        "high": high_spending
    }


# =========================================================
# 2. TEST DATA LOADING
# =========================================================

def test_data_loading_function():
    print("🧪 Testing data loading function...")

    datasets = create_test_datasets()

    # Save normal dataset
    datasets["normal"].to_csv("test_normal.csv", index=False)

    df = load_and_clean_transaction_data("test_normal.csv")

    # Core checks (robust, NOT fixed row count)
    assert len(df) > 0, "Should return data after cleaning"
    assert "amount" in df.columns, "Amount column missing"
    assert pd.api.types.is_float_dtype(df["amount"]), "Amount must be float"
    assert df["amount"].isna().sum() == 0, "No NaN values allowed"

    print("✔ Data loading tests passed")


# =========================================================
# 3. TEST SPENDING ANALYSIS
# =========================================================

def test_spending_analysis():
    print("🧪 Testing spending analysis...")

    df = pd.DataFrame({
        "date": pd.to_datetime(["2026-04-01", "2026-04-05", "2026-04-10"]),
        "amount": [100, 200, -50],
        "category": ["Food", "Transport", "Refund"],
        "description": ["A", "B", "C"]
    })

    df["day"] = df["date"].dt.day
    df["type"] = df["amount"].apply(lambda x: "refund" if x < 0 else "expense")
    df["amount_abs"] = df["amount"].abs()

    expenses = df[df["type"] == "expense"]

    total = expenses["amount_abs"].sum()

    # Refund excluded
    assert total == 300, "Refund should be excluded"

    # Category logic
    cat_totals = expenses.groupby("category")["amount_abs"].sum()

    assert cat_totals["Food"] == 100
    assert cat_totals["Transport"] == 200

    # Percent check
    assert round((cat_totals.sum() / total) * 100, 2) == 100.00

    print("✔ Spending analysis tests passed")


# =========================================================
# 4. TEST BUSINESS INSIGHTS
# =========================================================

def test_business_insights():
    print("🧪 Testing business insights...")

    df = pd.DataFrame({
        "amount": [300, 200, 100, 50],
        "category": ["Entertainment", "Food", "Entertainment", "Transport"]
    })

    df["amount_abs"] = df["amount"]

    cat_totals = df.groupby("category")["amount_abs"].sum()

    # Highest spending category
    assert cat_totals.idxmax() == "Entertainment"

    # Savings calculation
    savings = cat_totals["Entertainment"] * 0.25
    assert savings == 100.0

    # Logic check
    assert "Food" in cat_totals.index

    print("✔ Business insights tests passed")


# =========================================================
# 5. RUN ALL TESTS
# =========================================================

print("🔍 COMPREHENSIVE TESTING SUITE")
print("=" * 40)

try:
    create_test_datasets()
    test_data_loading_function()
    test_spending_analysis()
    test_business_insights()

    print("\n✅ ALL TESTS PASSED — FINANCE ASSISTANT IS WORKING")

except AssertionError as e:
    print(f"\n❌ TEST FAILED: {e}")

except Exception as e:
    print(f"\n⚠️ ERROR OCCURRED: {e}")

🔍 COMPREHENSIVE TESTING SUITE
🧪 Testing data loading function...
  Loaded 30 transactions
  Expenses : 27
  Refunds  : 3

      date  amount      category                              description  day    type  amount_abs
2026-04-01   18.50          Food    Maccas breakfast -Bourke St Melbourne    1 expense       18.50
2026-04-02    4.60     Transport        Myki top-up - Flinders St Station    2 expense        4.60
2026-04-03    6.80          Food      Flat white & muffin - Gloria Jean's    3 expense        6.80
2026-04-04   24.99 Entertainment             Netflix monthly subscription    4 expense       24.99
2026-04-05    4.60     Transport                Myki fare - bus route 246    5 expense        4.60
2026-04-06   11.90          Food            Subway footlong - Swanston St    6 expense       11.90
2026-04-07   59.99       Clothes                   Jeans - Cotton On City    7 expense       59.99
2026-04-08   22.50 Entertainment            Cinema ticket - Event Cinemas    8 expense

## Advanced Integration Tests

In [ ]:
# 🤖 AI Collaboration: Integration Testing
# Ask AI to help test the complete system integration

def test_full_workflow():
    """
    Test the complete workflow from CSV upload to final recommendations

    🤖 AI Collaboration Prompt:
    "Create an end-to-end test that:
    1. Loads sample CSV data
    2. Runs complete analysis pipeline
    3. Generates chat responses about the data
    4. Verifies RAG system retrieval
    5. Tests custom tool functionality
    Ensure all components work together seamlessly."
    """
    print("🧪 Testing complete workflow integration...")
    # Your AI-generated integration tests go here
    pass

def test_error_handling():
    """
    Test error handling and user experience

    🤖 AI Collaboration Prompt:
    "Create tests that verify error handling for:
    - Invalid file uploads
    - Network connection issues
    - Malformed data
    - User input validation
    Ensure error messages are user-friendly and helpful."
    """
    print("🧪 Testing error handling...")
    # Your AI-generated error tests go here
    pass

# Run integration tests
try:
    test_full_workflow()
    test_error_handling()
    print("✅ Integration tests completed successfully!")
except Exception as e:
    print(f"⚠️ Integration test issue: {e}")

---

# 📊 Project Completion Checklist

## Foundation Skills ✅
- [ ] **Data Processing**: CSV loading and cleaning functions work reliably
- [ ] **Analysis Functions**: Spending summaries calculate correctly
- [ ] **Business Insights**: Recommendations are relevant and actionable  
- [ ] **Error Handling**: Graceful handling of data issues
- [ ] **Testing**: Comprehensive test coverage for core functions
- [ ] **Documentation**: Clear AI collaboration documentation in diary

## Advanced Integration ✅
- [ ] **Chat Interface**: Finance advisor personality implemented
- [ ] **RAG System**: Document retrieval for financial guidance
- [ ] **Custom Tools**: At least one financial calculator/utility
- [ ] **Gradio UI**: Professional, user-friendly interface
- [ ] **Full Integration**: All components work together seamlessly

## Professional Standards ✅
- [ ] **Code Quality**: Professional, commented, maintainable code
- [ ] **Business Focus**: Clear connection to real finance problems
- [ ] **User Experience**: Interface suitable for non-technical users
- [ ] **AI Collaboration**: Extensive, well-documented AI usage
- [ ] **Testing**: Robust validation of all features

## Project Documentation ✅  
- [ ] **Developer's Diary**: Complete AI collaboration documentation
- [ ] **README**: Clear project description and usage instructions
- [ ] **GitHub**: Regular commits showing development progress
- [ ] **Reflection**: Thoughtful analysis of learning and challenges

---

# 🎯 Final Thoughts: Your Finance Assistant Journey

Congratulations on building your Smart Finance Assistant! This project represents a significant achievement in modern business programming:

**Technical Skills Developed:**
- AI-assisted development workflows
- Professional data processing with pandas
- Integration of multiple AI technologies
- User interface design with Gradio
- Comprehensive software testing

**Business Skills Developed:**  
- Financial data analysis and insights
- User-centered application design
- Professional documentation practices
- Iterative development methodology
- Critical evaluation of AI suggestions

**Professional Preparation:**
- Experience with industry-standard AI collaboration
- Portfolio-ready application development
- Understanding of business problem-solving with technology
- Documentation practices for workplace environments

**Your Smart Finance Assistant demonstrates your ability to direct AI tools toward meaningful business solutions - exactly the skill set that modern BIS graduates need for career success!**

---

*Remember to document all AI collaborations in your Developer's Diary and maintain regular GitHub commits throughout your development process.*
